fvm/src/kernel/mod.rs

```diff
@@ -48,36 +48,12 @@ pub struct CallResult {
     pub exit_code: ExitCode,
 }
 
-/// The "kernel" implements the FVM interface as presented to the actors. It:
-///
-/// - Manages the Actor's state.
-/// - Tracks and charges for IPLD & syscall-specific gas.
-///
-/// Actors may call into the kernel via the syscalls defined in the [`syscalls`][crate::syscalls]
-/// module.
-pub trait Kernel:
-    SyscallHandler<Self>
-    + ActorOps
-    + IpldBlockOps
-    + CircSupplyOps
-    + CryptoOps
-    + DebugOps
-    + EventOps
-    + GasOps
-    + MessageOps
-    + NetworkOps
-    + RandomnessOps
-    + SelfOps
-    + LimiterOps
-    + 'static
-{
-    /// The [`Kernel`]'s [`CallManager`] is
+/// The base [`Kernel`] trait implemented by all kernels.
+pub trait Kernel: SyscallHandler + 'static {
```

Comment:
I've also tried:

1. Inverting this to `SyscallHandler: Kernel`. I like that in theory, but it's strictly uglier in practice.
2. `SyscallHandler + Kernel` everywhere because they're _kind_ of orthogonal and only happen to be implemented on the same type. This actually helps a bit with some circular type issues so I may have to go back to it.

I haven't yet tried merging them into a single trait but.... then I'd need to define `Kernel` as `Kernel<K: Kernel>`, and I'd rather not do that.

I also haven't tried _completely_ separating them and treating them as entirely separate types. I might try doing that next to see if it helps with some of the type issues.

Reply:
Ok, I tried separating them and that's worse. I've moved it to an associated type and that seems like the best way forwards.